dash_cytoscape/utils/Tree.py

```diff
@@ -0,0 +1,158 @@
+from collections import deque
+
+
+class Tree(object):
+    def __init__(self, node_id, children=None, data=None, edge_data=None):
+        """
+        A class to facilitate tree manipulation in Cytoscape
+        :param node_id: The ID of this tree, passed to the node data dict
+        :param children: The children of this tree, also Tree objects
+        :param data: Dictionary passed to this tree's node data dict
+        :param edge_data: Dictionary passed to the data dict of the edge connecting this tree to its
+        parent
+        """
+        self.node_id = node_id
+        self.edge_data = edge_data
+        self.data = data
+        self.children = children
+
+        self.index = {}
+
+        if not children:
+            self.children = []
+        if not data:
+            self.data = {}
+        if not edge_data:
+            self.edge_data = {}
+
+    def _dfs(self, search_id):
+        if self.node_id == search_id:
+            return self
+        elif self.is_leaf():
+            return None
+        else:
+            for child in self.children:
+                result = child.dfs()
+                if result:
+                    return result
+
+            return None
+
+    def _bfs(self, search_id):
+        stack = deque([self])
+
+        while stack:
+            tree = stack.popleft()
+
+            if tree.node_id == search_id:
+                return tree
+
+            if not tree.is_leaf():
+                for child in tree.children:
+                    stack.append(child)
+
+        return None
+
+    def is_leaf(self):
+        """
+        :return: If the Tree is a leaf or not
+        """
+        return not self.children
+
+    def add_child(self, child):
+        """
+        Add a single child to the children of a Tree
+        :param child: Tree object
+        """
+        self.children.append(child)
+
+    def add_children(self, children):
+        """
+        Add a list of children to the current children of a Tree
+        :param children: List of Tree objects
+        """
+        self.children.extend(children)
```

Comment:
You may want to `assert len(children) > 1`.
Is `add_child` a special case of `add_children`, which would still work when `len(children) == 1`?

Reply:
Yes. Add child is a special case of add_children. Sometimes the programmer might not know in advance how many children the tree might have (when it's decided programmatically). So they might just use add_children() with a list all the time.

Reply:
@xhlulu Then I would recommend removing method `add_child` and keep only method `add_children` (so there's *one* way to add children, however many they end up being).